r_code/synthetic_dgp.R

```diff
@@ -0,0 +1,1027 @@
+                         num_factors = 3,
+                         loading_scale = 0,
+                         intercept_scale = 0,
+                         conditional_impact_het = -0,
+                         rescale_y_mean = 2.5e4,
+                         seed = 19) {
+
+  # Match the arguments to valid entries.
+  set.seed(seed)
+  if (missing(selection)) {
+    selection <- "random"
+  } else {
+    selection <- match.arg(selection)
+  }
+
+  if (missing(freq)) {
+    freq <- "monthly"
+  } else {
+    freq <- match.arg(freq)
+  }
+  # Require 5% min in both treat and control.
+  stopifnot(prop_treated >= 0.05 & prop_treated <= 0.95)
+  # Require cov_overlap_scale to be between -1 (shift treat down) and 1.
+  stopifnot(cov_overlap_scale <= 1 & cov_overlap_scale >= -1)
+  # Require 3 factors of more for monthly, 4 for weekly, and 5 daily.
+  if (freq == "monthly") stopifnot(num_factors >= 3)
+  if (freq == "weekly") stopifnot(num_factors >= 4)
+  if (freq == "daily") stopifnot(num_factors >= 5)
+  # Require less than 100% TE.
+  stopifnot(treat_impact_mean < 1)
+  # Require ACF to less than one, and not too negative.
+  stopifnot(rho < 1 & rho > -0.25)
+  stopifnot(rho * rho_shift < 1 & rho * rho_shift > -0.25)
+  # Require the intercept scale to be between -1 and 1.
+  stopifnot(intercept_scale <= 1 & intercept_scale >= -1)
+  # Require the intercept scale to be between -1 and 1.
+  stopifnot(loading_scale <= 1 & loading_scale >= -1)
+  # Stop if there are too few pre treat periods, 20 is a rule of thumb.
+  stopifnot(first_treat < 0.8 * num_periods | first_treat >= 20)
+
+
+  # Assign covariates and treatment given selection and overlap.
+  synth_data_unit <- .UnitLevelSimulation(
+    n_inp = num_entries,
+    type_inp = selection,
+    cov_overlap_inp = cov_overlap_scale,
+    loading_scale_inp = loading_scale,
+    num_factors_inp = num_factors,
+    int_scale_inp = intercept_scale,
+    rho_inp = rho,
+    rho_scale_inp = rho_scale,
+    rho_shift_inp = rho_shift,
+    prop_treated_inp = prop_treated,
+    treat_start = first_treat,
+    num_periods_inp = num_periods,
+    impact_mean_inp = treat_impact_mean,
+    impact_sd_inp = treat_impact_sd,
+    decay_mean_inp = treat_decay_mean,
+    decay_sd_inp = treat_decay_sd
+  )
+
+  # Work on time varying characteristics.
+  synth_data_factors <- .GenerateFactors(
+    num_factors_inp = num_factors,
+    num_periods_inp = num_periods,
+    freq_inp = freq
+  )
+  # Combine time and unit characteristics into a grid.
+  unit_time_grid <- tidyr::expand_grid(
+    entry = seq_len(num_entries),
+    time = seq_len(num_periods)
+  )
+
+  synth_data_full <- unit_time_grid %>%
+    dplyr::left_join(synth_data_unit, by = c("entry")) %>%
+    dplyr::left_join(synth_data_factors, by = c("time"))
+
+  rescale_y <- dplyr::case_when(
+    freq == "monthly" ~ log(rescale_y_mean),
+    freq == "weekly" ~ log(rescale_y_mean / 4),
+    freq == "daily" ~ log(rescale_y_mean / 30)
```

Comment:
Not in any significant way. This is just a way to rescale the outcome so that it doesn't have massive outliers (rescaley_y_mean helps determine the mean). So weekly generated data might not exactly add up to monthly generated data when aggregated, but that doesn't actually impact the estimation (which is what matters).